File: frontend/src/types/homework.ts
/**
 * 作业类型定义 — V2 组卷版
 *
 * 功能：定义组卷页所需的全部数据类型，包括作业基本信息、题目项、页面配置
 * 输入参数：无（仅类型）
 * 返回值：导出 TypeScript 类型
 * 使用场景：组卷页、作业列表、PDF 导出等
 */

/** 作业中的题目项 */
export interface HomeworkQuestionItem {
  id: string
  question_id: string
  sort_order: number
  score: number
  is_required: boolean
  stem: string
  question_type: string
  question_no: number
  options: Array<{ label?: string; text?: string; content?: string }> | string[]
  answer: string | null
}

/**
 * 作业页面配置（页眉/页脚/水印/logo 等）
 *
 * 新增字段（阶段4：页眉/Logo 独立可拖拽）：
 *  - header_box / logo_box: 元素在画布中的位置与大小
 *    坐标系：相对画布纸张左上角，单位 px（已含 scale 缩放）
 *    x/y: 元素左上角坐标
 *    width/height: 元素宽高
 *    show: 是否显示（不填则视为 true）
 */
export interface HomeworkPageConfig {
  paper_size?: "A3" | "A4"
  header_text?: string
  header_font_size?: number
  /** 页眉文字的位置与大小（独立可拖拽元素） */
  header_box?: ElementBox
  footer_text?: string
  footer_font_size?: number
  watermark_text?: string
  watermark_opacity?: number
  watermark_angle?: number
  watermark_size?: number
  logo_url?: string
  logo_width?: number
  /** Logo 位置与大小（独立可拖拽 + 等比缩放） */
  logo_box?: ElementBox
  /** 试卷标题的位置与大小（独立可拖拽元素，可在 A4 画布上自由移动） */
  title_box?: ElementBox
  question_font_size?: number
  title_font_size?: number
  info_font_size?: number
  show_subject_grade?: boolean
  show_name_class?: boolean
  /** 需求 3：每道题用户手动添加的留白行数映射（key = homework_questions.id） */
  blank_lines?: Record<string, number>
}

/**
 * 元素在画布中的位置与大小
 *
 * 输入参数：
 *  - x/y - 左上角坐标（相对画布纸张，单位 px）
 *  - width/height - 元素宽高
 *  - z_index - 图层层级（数值越大越靠上，相同则按 logo_box < header_box 顺序）
 *  - show - 是否显示（默认 true）
 *  - locked - 是否锁定（默认 false，锁定后不可拖动/缩放）
 * 使用场景：Photoshop 式图层系统的页眉/Logo 独立可拖拽元素
 */
export interface ElementBox {
  x: number
  y: number
  width: number
  height: number
  /** 图层层级（数值越大越靠上层）；不填则按默认顺序：标题<页眉<水印<Logo */
  z_index?: number
  /** 是否显示（默认 true） */
  show?: boolean
  /** 是否锁定（默认 false）；锁定后不可拖动/缩放/编辑 */
  locked?: boolean
}

/** 作业 */
export interface Homework {
  id: string
  title: string
  subject: string | null
  grade: string | null
  total_score: number
  status: string
  page_config: HomeworkPageConfig | null
  created_at: string | null
  updated_at: string | null
  questions: HomeworkQuestionItem[]
}

File: backend/app/schemas/homework.py
"""
作业 Schema

功能：定义作业相关的请求/响应数据结构
输入参数：无（Pydantic 模型定义）
返回值：HomeworkQuestionItem / HomeworkResponse / HomeworkListResponse 等类
使用场景：作业创建、编辑、查询、题目排序
"""
from pydantic import BaseModel
from datetime import datetime


class HomeworkQuestionItem(BaseModel):
    """
    作业题目项

    功能：定义作业中单个题目的信息
    输入参数：题目关联信息和题目内容
    返回值：作业题目项对象
    使用场景：作业详情中的题目列表
    """
    id: str  # 关联ID
    question_id: str  # 题目ID
    sort_order: int  # 排序序号
    score: int  # 分值
    is_required: bool  # 是否必做
    stem: str = ""  # 题干
    question_type: str = "general"  # 题型
    question_no: int = 0  # 题号
    options: list = []  # 选项列表
    answer: str | None = None  # 答案

    model_config = {"from_attributes": True}  # 支持从ORM对象转换


class HomeworkResponse(BaseModel):
    """
    作业响应

    功能：定义作业详情的返回格式
    输入参数：作业各字段
    返回值：作业响应对象
    使用场景：作业详情查询
    """
    id: str  # 作业ID
    title: str  # 作业标题
    subject: str | None = None  # 学科
    grade: str | None = None  # 年级
    total_score: int = 0  # 总分
    status: str = "draft"  # 状态
    page_config: dict = {}  # 页面配置
    created_at: datetime  # 创建时间
    updated_at: datetime  # 更新时间
    questions: list[HomeworkQuestionItem] = []  # 题目列表

    model_config = {"from_attributes": True}  # 支持从ORM对象转换


class HomeworkListResponse(BaseModel):
    """
    作业列表响应

    功能：定义作业列表的返回格式
    输入参数：homework（作业列表）、total（总数）
    返回值：作业列表响应对象
    使用场景：作业列表查询接口
    """
    homework: list[HomeworkResponse]  # 作业列表
    total: int  # 总数


class HomeworkCreateRequest(BaseModel):
    """
    作业创建请求

    功能：定义创建作业的请求参数
    输入参数：title（作业标题，可选）、paper_id（来源试卷ID，可选）、
             subject/grade（学科年级）、page_config（页面配置，可选）
    返回值：作业创建请求对象
    使用场景：从试卷或独立创建作业
    """
    title: str | None = None  # 作业标题（直接创建时可指定）
    paper_id: str | None = None  # 来源试卷ID（从试卷创建时必填）
    subject: str | None = None  # 学科
    grade: str | None = None  # 年级
    page_config: dict | None = None  # 页面配置


class HomeworkUpdateRequest(BaseModel):
    """
    作业更新请求

    功能：定义更新作业的请求参数
    输入参数：title、subject、grade、page_config 等可选字段
    返回值：作业更新请求对象
    使用场景：编辑作业信息
    """
    title: str | None = None  # 作业标题
    subject: str | None = None  # 学科
    grade: str | None = None  # 年级
    total_score: int | None = None  # 总分
    page_config: dict | None = None  # 页面配置


class AddQuestionRequest(BaseModel):
    """
    添加题目请求

    功能：定义向作业添加题目的请求参数
    输入参数：question_id、score
    返回值：添加题目请求对象
    使用场景：向作业中添加题目
    """
    question_id: str  # 题目ID
    score: int = 0  # 分值


class ReorderRequest(BaseModel):
    """
    题目排序请求

    功能：定义作业题目重新排序的请求参数
    输入参数：question_ids（按新顺序排列的ID列表）
    返回值：排序请求对象
    使用场景：拖拽排序作业题目
    """
    question_ids: list[str]  # 按新顺序排列的 homework_question id 列表


class SetScoreRequest(BaseModel):
    """
    设置分值请求

    功能：定义设置题目分值的请求参数
    输入参数：score（分值）
    返回值：设置分值请求对象
    使用场景：修改作业中某题的分值
    """
    score: int  # 分值
